Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp

```diff
@@ -0,0 +1,370 @@
+    : m_input(WTFMove(input))
+{
+}
+
+// https://urlpattern.spec.whatwg.org/#rewind
+void URLPatternConstructorStringParser::rewind()
+{
+    m_tokenIndex = m_componentStart;
+    m_tokenIncrement = 0;
+}
+
+// https://urlpattern.spec.whatwg.org/#get-a-safe-token
+const URLPatternUtilities::Token& URLPatternConstructorStringParser::getSafeToken(size_t index) const
+{
+    if (index < m_tokenList.size())
+        return m_tokenList[index];
+
+    ASSERT(m_tokenList.size() >= 1);
+    ASSERT(m_tokenList.last().type == URLPatternUtilities::TokenType::End);
+    return m_tokenList.last();
+}
+
+// https://urlpattern.spec.whatwg.org/#is-a-non-special-pattern-char
+bool URLPatternConstructorStringParser::isNonSpecialPatternChar(size_t index, char value) const
+{
+    auto token = getSafeToken(index);
+    // FIXME: Find a better way to convert a char to a string.
+    StringView fixme = StringView(std::span<const char>(&value, 1));
+
+    return token.value == fixme.toString()
+        && (token.type == URLPatternUtilities::TokenType::Char
+        || token.type == URLPatternUtilities::TokenType::EscapedChar
+        || token.type == URLPatternUtilities::TokenType::InvalidChar);
+}
+
+// https://urlpattern.spec.whatwg.org/#is-a-search-prefix
+bool URLPatternConstructorStringParser::isSearchPrefix() const
+{
+    if (isNonSpecialPatternChar(m_tokenIndex, '?'))
+        return true;
+    if (m_tokenList[m_tokenIndex].value != "?"_s)
+        return false;
+
+    auto previousIndex = m_tokenIndex - 1;
+    if (previousIndex < 0)
+        return true;
+
+    auto previousToken = getSafeToken(previousIndex);
+    if ( previousToken.type == URLPatternUtilities::TokenType::Name
+        || previousToken.type == URLPatternUtilities::TokenType::Regexp
+        || previousToken.type == URLPatternUtilities::TokenType::Close
+        || previousToken.type == URLPatternUtilities::TokenType::Asterisk )
+        return false;
+    return true;
+}
+
+// https://urlpattern.spec.whatwg.org/#next-is-authority-slashes
+bool URLPatternConstructorStringParser::isAuthoritySlashesNext() const
+{
+    if (!isNonSpecialPatternChar(m_tokenIndex+1, '/'))
+        return false;
+    if (!isNonSpecialPatternChar(m_tokenIndex+2, '/'))
+        return false;
+    return true;
+}
+
+// https://urlpattern.spec.whatwg.org/#make-a-component-string
+String URLPatternConstructorStringParser::makeComponentString() const
+{
+    ASSERT(m_tokenIndex < m_tokenList.size());
+    const auto& token = m_tokenList[m_tokenIndex];
+
+    auto componentStartToken = getSafeToken(m_componentStart);
+    ASSERT(componentStartToken.index && token.index);
+    auto componentStartIndex = *componentStartToken.index;
+
+    return m_input.substring(componentStartIndex, *token.index - componentStartIndex).toString();
+}
+
+static inline void setInitComponentFromState(URLPatternInit& init, URLPatternConstructorStringParserState state, String componentString)
```

Comment:
`String` should never be passed by value.

Reply:
I know this and this was a careless mistake on my part 😓. 

I will change to String&&.